source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,340 @@
+    } else if (string_header_name.at(0) == ':' &&
+               !kAllowedPseudoHeaders.contains(header_name.getStringView())) {
+      // This is an unrecognized pseudo header, reject the response
+      reject_details = UhvResponseCodeDetail::get().InvalidPseudoHeader;
+    } else {
+      auto entry_result = validateResponseHeaderEntry(header_name, header_value);
+      if (!entry_result) {
+        reject_details = static_cast<std::string>(entry_result.details());
+      }
+    }
+
+    return reject_details.empty() ? ::Envoy::Http::HeaderMap::Iterate::Continue
+                                  : ::Envoy::Http::HeaderMap::Iterate::Break;
+  });
+
+  return reject_details.empty()
+             ? ResponseHeaderMapValidationResult::success()
+             : ResponseHeaderMapValidationResult(RejectAction::Reject, reject_details);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateTEHeader(const ::Envoy::Http::HeaderString& value) {
+  //
+  // Only allow a TE value of "trailers" for HTTP/2, based on
+  // RFC 7540, https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2:
+  //
+  // The only exception to this is the TE header field, which MAY be present
+  // in an HTTP/2 request; when it is, it MUST NOT contain any value other
+  // than "trailers".
+  //
+  if (!absl::EqualsIgnoreCase(value.getStringView(), header_values_.TEValues.Trailers)) {
+    return {RejectAction::Reject, Http2ResponseCodeDetail::get().InvalidTE};
+  }
+
+  return HeaderEntryValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateAuthorityHeader(const ::Envoy::Http::HeaderString& value) {
+  //
+  // From RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.2:
+  //
+  // authority = [ userinfo "@" ] host [ ":" port ]
+  //
+  // HTTP/2 deprecates the userinfo portion of the :authority header. Validate
+  // the :authority header and reject the value if the userinfo is present. This
+  // is based on RFC 7540, https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.3
+  //
+  // The host portion can be any valid URI host, which this function does not
+  // validate. The port, if present, is validated as a valid uint16_t port.
+  //
+  return validateHostHeader(value);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateGenericHeaderName(const ::Envoy::Http::HeaderString& key) {
+  //
+  // For HTTP/2, connection-specific headers must be treated as malformed. From RFC 7540,
+  // https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2:
+  //
+  // any message containing connection-specific header fields MUST be treated
+  // as malformed (Section 8.1.2.6).
+  //
+  static const absl::node_hash_set<absl::string_view> kRejectHeaderNames = {
+      "transfer-encoding", "connection", "upgrade", "keep-alive", "proxy-connection"};
```

Comment:
I think `transfer-encoding: trailers` is allowed in H/2 https://www.rfc-editor.org/rfc/rfc7540.html#section-8.1.2.2

Reply:
I'm slightly confused on the difference between the `TE` header and the `transfer-encoding`. My understanding for H2 is that `TE` is allowed but can only be `trailers` and `transfer-encoding` is never allowed because it is a connection header. I just want to make sure I understand the difference before i make a change here.

Reply:
Sorry this is my confusion. Yes, it is `TE` header